### DISCUSSION TITLE
Live Preview option to automatically add a time-limited token to the query (incl. ENV vars)

### DISCUSSION THREAD

**Discussion by formfcw** [directus/directus#20302] [Feature Requests]
### Summary

_Provide an easy way to (live) preview apps/sites that use unpublished and protected data._

In the data model settings of a collection we can specify a `Preview URL`. It would be nice if we could add a temporary authentication token to this URL, so that the front-end application can access unpublished data via the API with a time-limited access token.

In addition, it would be nice if we could set the base URL for the `Preview URL` as an ENV variable (like `PROJECT_URL`), so that when a relative Preview URL is specified, this base URL (`PROJECT_URL`) will be taken.

### Basic Example

I have already created an endpoint extension that does just that: [Tokenized Preview Endpoint](https://github.com/formfcw/directus-extension-tokenized-preview).

### Motivation

The ability to attach a time-limited token to the `Preview URL` allows a front-end application to access content/data that is not publicly accessible via the API. This is basically the main reason why I want to preview an item.

Setting the base URL as an ENV variable, as described above, is useful if we are deploying the application(s) in a different environment.

### Detailed Design

In addition to specifying the `Preview URL` in the collection’s data model settings, we should also have an option to append a temporary access token. This could be a checkbox below the `Preview URL` field, or it could be built into the `Preview URL` field itself.

This will turn a URL from …

`http://localhost:3000/posts/{{ slug }}?preview=true`
… or …
`http://localhost:3000/posts/{{ slug }}?preview=true&token={{ token }}`

… to something like:

`http://localhost:3000/posts/{{ slug }}?preview=true&access_token=q90357o234afi4309…2jsdfa2908287`

If it’s solved with a checkbox, we should provide an ENV variable to set the token key, such as `PREVIEW_TOKEN_KEY=access_token`.

The token should be time-limited, but valid as long as the (Directus) user is logged in (into Directus).

To make it more portable (to another environment), as described above, it would be useful if we could use relative `Preview URL`s that use our project URL (front-end) as the base URL.

So if we have a `Preview URL` like `/posts/{{ slug }}?preview=true` and set an ENV variable like `PROJECT_URL` or `PREVIEW_BASE_URL` with a value of `http://localhost:3000`, the result should be:

`http://localhost:3000/posts/{{ slug }}?preview=true` (example without token)

### Must Have:
- Ability to append a token to the `Preview URL`
- Ability to set the base URL as an ENV variable

### Should Have:
- an interface (checkbox or the `Preview URL` field itself) for attaching a token

### Could Have:
- ENV var to specify the name of the access token key used in the URL like `&access_token=…` or `&token=…` or `&tk=…`

### Drawbacks

No disadvantages that I can think of.

### Alternatives

An alternative would be to append a token string to each `Preview URL`, by setting an(other) ENV var like `TOKENIZED_PREVIEW=true`.

### Adoption Strategy

As this is a new feature that must be actively enabled, it has no effect on existing instances.

**Comment by arladmin** [comment]
I would also suggest the option to (must) have `token = 'the current access token of the logged-in user' `.

**Comment by formfcw** [comment]
Since Directus `^10.10.0` supports session cookies for auth, we can use this token (`directus_session_token`) directly and automatically add it as access token to the preview url. I have now adapted the latter to my extension ([Tokenized Preview](https://github.com/formfcw/directus-extension-tokenized-preview)).
